packages/dds/tree/src/feature-libraries/sequence-change-family/rebase.ts

```diff
@@ -0,0 +1,116 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    ChangesetTag,
+    getMarkLength,
+    isAttachGroup,
+    isReattach,
+    isSkipMark,
+    MarkListFactory,
+    splitMark,
+    Transposed as T,
+} from "../../changeset";
+import { clone, fail } from "../../util";
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Rebases `change` over `base` assuming they both apply to the same initial state.
+ * @param change - The changeset to rebase.
+ * @param base - The changeset to rebase over.
+ * @returns A changeset that performs the changes in `change` but does so assuming `base` has been applied first.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Marks that affect existing content are removed instead of muted when rebased over the deletion of that content.
+ *   This prevents us from then reinstating the mark when rebasing over the revive.
+ * - Tombs are not added when rebasing an insert over a gap that is immediately left of deleted content.
+ *   This prevents us from being able to accurately track the position of the insert.
+ * - Tiebreak ordering is not respected.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function rebase(change: SequenceChangeset, base: SequenceChangeset): SequenceChangeset {
+    const fields = rebaseFieldMarks(change.marks, base.marks);
+    return {
+        marks: fields,
+    };
+}
+
+function rebaseFieldMarks(change: T.FieldMarks, base: T.FieldMarks): T.FieldMarks {
+    const fields: T.FieldMarks = {};
+    for (const key of Object.keys(change)) {
+        if (key in base) {
+            fields[key] = rebaseMarkList(change[key], base[key]);
+        } else {
+            fields[key] = clone(change[key]);
+        }
+    }
+    return fields;
+}
+
+function rebaseMarkList(currMarkList: T.MarkList, baseMarkList: T.MarkList): T.MarkList {
+    const factory = new MarkListFactory();
+    let iBase = 0;
+    let iCurr = 0;
+    let nextCurrMark: T.Mark | undefined = currMarkList[iCurr];
+    let nextBaseMark: T.Mark | undefined = baseMarkList[iBase];
+    while (nextCurrMark !== undefined && nextBaseMark !== undefined) {
+        let currMark: T.Mark = nextCurrMark;
+        let baseMark: T.Mark = nextBaseMark;
+        nextCurrMark = undefined;
+        nextBaseMark = undefined;
+
+        if (isAttachGroup(currMark) || isReattach(currMark)) {
```

Comment:
Assuming reattaches shouldn't be part of attach groups, we might want a helper that checks for these two cases.

Reply:
These are now in separate ifs to allow for better commenting